Changes from Unofficial Mator Smash Updated

chuckseven1 opened the pull request:
Please tell me if any of these changes look insane. I made a few... educated guesses when changing from the xEdit 3 API to 4.

Commit:
Add missing bash tag settings

Commit:
Update to xedit 4.0.3

Commit:
Fix handling of integers which hold flags

Commit:
Fix handling of arrays

Commit:
Fix up bash tag settings now that flags work

Commit:
Delete hardcoded dat files

No longer needed with newer xEdit

Commit:
Increment to v1.1.0

Commit:
Support for loading more than 255 plugins w/ESL

Commit:
Fix for Skyrim VR

Still treats it the same as SSE, but should detect it properly now.

Commit:
Load Creation Club mods before plugins.txt mods

Commit:
Attempt to fix bug in smashing algo I introduced

matortheeternal commented:
Alright, I'll review this over the next few days.  :+1:

Commit:
Improve limiting of masters

Should help to smash more plugins without reaching 253 master limit.

matortheeternal commented on frontend/msSmash.pas:
```diff
@@ -306,7 +303,8 @@ procedure SmashRecords(var patch: TPatch; var records: TInterfaceList);
           e := WinningOverrideInFiles(rec, patch.plugins);
         Tracker.Write(Format('  [%d] Copying record %s', [i + 1, e.Name]));
-        eCopy := wbCopyElementToFile(e, patchFile, false, true, '', '' ,'');
+        AddRequiredMaster(patch, e._File);
```
this isn't what you should be doing.  instead do this:
<https://github.com/TES5Edit/TES5Edit/blob/dev-4.1.4/xEdit/xeMainForm.pas#L1884-L1923>
(you can't access this function, so just rewrite `AddRequiredMaster` based on it, and rename to `AddRequiredMasters`.)

The key here is using `ReportRequiredMasters` to get the masters required by the element and then adding those masters to the patch plugin.

chuckseven1 replied:
I'm dumb. Thanks for pointing this out.

matortheeternal commented on frontend/msSmash.pas:
```diff
@@ -334,6 +332,7 @@ procedure SmashRecords(var patch: TPatch; var records: TInterfaceList);
           mst := WinningOverrideInFiles(rec, plugin.masters);
         Tracker.Write(Format('    Smashing override from: %s, master: %s',
           [f.FileName, mst._File.FileName]));
+        AddRequiredMaster(patch, rec._File);
```
same thing here as previous comment

matortheeternal commented on frontend/msSmash.pas:
```diff
@@ -461,10 +460,12 @@ procedure CleanPatch(var patch: TPatch);
+    // remove ITPOs
     if not settings.preserveITPOs then
       RemoveITPOs(patchFile);
+    patchFIle.CleanMasters;
```
at this stage you don't need to `CleanMasters`, so don't bother with it.  just because a master can be removed doesn't mean it should be removed.  I prefer to have every plugin that could have influenced the generation of the patch plugin as a master for the plugin, as it strictly communicates "if you remove this plugin from your load order, you have to regenerate the patch".

chuckseven1 replied:
I still have the plugin metadata remember all the plugins that were used in the smashing even if they are not a master, so the UI will tell you to rebuild the patch if you remove one whether it was a master or not. I can remove this if you prefer though.

matortheeternal replied:
either way.  because we're using `AddRequiredMasters` anyways my point is kind of moot, so you can leave this I suppose.

matortheeternal commented on lib/mte/mteBase.pas:
```diff
@@ -200,16 +200,10 @@ class function TPluginHelpers.CreateNewBasePlugin(var list: TList; filename: str
   // create new plugin file
   SysUtils.FormatSettings.DecimalSeparator := '.';
-  aFile := wbNewFile(wbDataPath + filename, LoadOrder);
+  aFile := wbNewFile(wbDataPath + filename, LoadOrder, false);
```
`LoadOrder` isn't set here?

chuckseven1 replied:
Looks like I deleted one line too many. Oops

matortheeternal commented on lib/mte/mteBase.pas:
```diff
@@ -1395,10 +1389,20 @@ procedure BuildChildDefs(obj: ISuperObject; def: IwbNamedDef);
   if Supports(def, IwbSubRecordDef, subDef) then
     BuildChildDefs(obj, subDef.GetValue as IwbNamedDef)
+  // try IwbFlagsDef
+  // TODO: Is this nil bad?? It seems to work...
```
pretty sure it's fine, I believe the idea here is that the formatter could change based on the element, so to know the correct formatter you actually technically need the element, but this is only the case on a few records that use format unions, and in those cases if you don't provide the element it'll just return the first entry in the union, and in practice all of these situations have every entry in the union as IwbFlagsDef.  It does mean that we may not display the correct flags in certain circumstances, for a few specific record types.

matortheeternal reviewed (changes requested):
A few things that need to be fixed.

matortheeternal commented:
Also it seems like string file loading may be broken based on the comments on Nexus Mods.  You should look into that.

chuckseven1 commented:
> Also it seems like string file loading may be broken based on the comments on Nexus Mods. You should look into that.

Yes I saw the comments on nexus. I have started looking into it. Not sure why it only seems to affect some localized strings, but I am clearly doing something wrong with them.

Commit:
Make changes suggested by mator

Commit:
Try to fix loadorder for VR

matortheeternal commented on frontend/msSmash.pas:
```diff
@@ -90,41 +90,62 @@ function GetPatchFile(var patch: TPatch; var lst: TList): IwbFile;
+  slMasters.Free;
+  if Tracker.Cancel then
+    raise Exception.Create('User cancelled smashing.');
+  Tracker.Write('Done adding masters');
```
you still need these lines of code to be in a `finally` block, else `slMasters` will be a memory leak when an exception happens.  The `Done adding masters` message can be moved up to the try block though.

matortheeternal commented on lib/mte/mteBase.pas:
```diff
@@ -200,6 +200,7 @@ class function TPluginHelpers.CreateNewBasePlugin(var list: TList; filename: str
   plugin: TBasePlugin;
 begin
   Result := nil;
+  LoadOrder := PluginsList.Count + 1;
```
I'm pretty sure `PluginsList.Count` should already be at the correct ordinal because `LoadOrder` is 0-based, so you shouldn't need to `+ 1`.

chuckseven1 replied:
I just restored that line from before my changes, the `+ 1` was there originally.

matortheeternal replied:
huh, ok.  might have been a bug then.  I'd have to look closer to see if that is indeed the case or if something else is going on.  it's fine for now I suppose :+1:

matortheeternal reviewed (changes requested):
two minor issues

matortheeternal commented:
> Yes I saw the comments on nexus. I have started looking into it. Not sure why it only seems to affect some localized strings, but I am clearly doing something wrong with them.

I think something in how localization is handled got changed from xEdit 3.x to 4.x, look into how they're doing localization shenanigans in the xEdit codebase and see if you can update things here appropriately.

If you can't see an issue you might have to get someone to send you their strings file(s).  It could be a lot of different things, might be difficult to debug.  You're certain it's only happening with *some* localized strings and not all?

chuckseven1 commented:
> > Yes I saw the comments on nexus. I have started looking into it. Not sure why it only seems to affect some localized strings, but I am clearly doing something wrong with them.
> 
> I think something in how localization is handled got changed from xEdit 3.x to 4.x, look into how they're doing localization shenanigans in the xEdit codebase and see if you can update things here appropriately.
> 
> If you can't see an issue you might have to get someone to send you their strings file(s). It could be a lot of different things, might be difficult to debug. You're certain it's only happening with _some_ localized strings and not all?

I had a string result in that error one time I smashed, then after changing my LO the same record was still in my smashed patch but that time it had the right value...

Yes I will look at xEdit to try to figure it out when I get the chance.

Commit:
Fix try/except/finally block

matortheeternal commented:
> I had a string result in that error one time I smashed, then after changing my LO the same record was still in my smashed patch but that time it had the right value...
> 
> Yes I will look at xEdit to try to figure it out when I get the chance.

OK, will merge and publish as soon as this is resolved.

Commit:
Turn off translation of localized strings

No need to translate them since we don't show them in the smash UI.
This seems to resolve the "No strings file for lstring" error.

Commit:
Log adding masters instead of printing to tracker

Now that masters get added dynamically, it made the tracker not useful.

matortheeternal commented on frontend/msLoader.pas:
```diff
@@ -106,6 +106,7 @@ function InitBase: boolean;
   wbEditAllowed := True;
   wbContainerHandler := wbCreateContainerHandler;
   wbContainerHandler._AddRef;
+  wbLocalizationHandler.NoTranslate := true;
```
I don't think you can do this, because when you copy a localized string in xEdit it copies the localized string value.  A localized string is basically just a `uint32` (`cardinal` in Delphi-land) which points to a specific id in the `strings` file associated with the ESP being processed.  For that reason, when copying a localized string it makes sense to copy not the `id` but the localized string value itself, because the `id` isn't meaningful outside of whatever plugin the record is getting copied from.

But I could also be misinterpreting what this option does?  Mostly saying this based on your commit comment.

chuckseven1 replied:
Yeah you are right. That setting does not do what I thought I did. It makes the errors go away but the strings are still wrong.

matortheeternal commented on frontend/msSmash.pas:
```diff
@@ -134,8 +133,8 @@ procedure AddRequiredMasters(var aFile: IwbFile; const el: IwbElement);
           aFile.CleanMasters;
 
         aFile.AddMasters(slMasters);
+        Logger.Write('PATCH', 'MASTERS', 'Added masters: ' + slMasters.CommaText);
```
:+1:

This is nice because it'll allow us to keep better track of the masters that are added during the smashing process.  An issue was reported to me the other day where the resulting smashed patch was missing a required master, so hopefully this can help debug that.

Commit:
Load plugin BSAs ourselves

Seems to fixed localized strings
Must be configuring xEdit wrong if it is not loading them?

Commit:
Add extra master for SkyrimVR

chuckseven1 commented:
The localized strings issue is because the BSAs corresponding to plugins in the LO are not being loaded. I added code to load them myself and the strings all come out correct for me. I must still be configuring xEdit wrong? I imagine it supports loading the BSAs automagically for the plugins.

matortheeternal commented:
> The localized strings issue is because the BSAs corresponding to plugins in the LO are not being loaded. I added code to load them myself and the strings all come out correct for me. I must still be configuring xEdit wrong? I imagine it supports loading the BSAs automagically for the plugins.

The BSA loading happens in msLoader.pas, LoadBSAs function.

chuckseven1 commented:
> > The localized strings issue is because the BSAs corresponding to plugins in the LO are not being loaded. I added code to load them myself and the strings all come out correct for me. I must still be configuring xEdit wrong? I imagine it supports loading the BSAs automagically for the plugins.
> 
> The BSA loading happens in msLoader.pas, LoadBSAs function.

Yes but that function was only loading the BSAs listed in skyrim.ini. I am talking about loading foo.bsa when foo.esp is loaded. That was not happening.

I added a check to load the BSAs associated with the plugins and now my localized strings come out right. The problem was I have localized plugins which had their strings file in their associated BSA. That is why the files were not being found.

matortheeternal commented:
> Yes but that function was only loading the BSAs listed in skyrim.ini. I am talking about loading foo.bsa when foo.esp is loaded. That was not happening.
> 
> I added a check to load the BSAs associated with the plugins and now my localized strings come out right. The problem was I have localized plugins which had their strings file in their associated BSA. That is why the files were not being found.

It definitely has code that loaded BSAs associated with plugin files that worked in the past.  If it isn't working in the current version that means the functions it calls may have changed, or something.

See these lines: https://github.com/matortheeternal/smash/blob/master/frontend/msLoader.pas#L291-L298

You should probably fix this function instead of adding a new one (haven't looked at your changes yet, about to run out of the house)

Sandman534 committed:
Update msSmash.pas

Add the required masters of the elements parent plugin. This will fix the "can not be mapped to file FileID for file" error.

Sandman534 committed:
Update mteBase.pas

When creating new settings this will make sure that Structs and Struct Arrays do not get prefixed with a signature

Commit:
Regenerate Smash.All setting

Fixed some issues I was having with null references in smashed patch.
TODO: Check with mator that description of Smash.All was true...

Commit:
Update bash tag settings

matortheeternal commented:
Alright, want to check in.  What's our current status here?  Is everything functioning correctly?  Localization issue is resolved?

chuckseven1 commented:
The localized strings are working, but with my added line. I found the old code that was supposed to load the plugin BSAs but have not yet figured out how it used to work.

Also I tried to regenerate the Smash.All setting because I was getting null references in my smashed patch due to slight changes in how xEdit named things (see c82f7fc58f24ac0fc45513f1f2a246970d09035f). It seems to work for me but other users are reporting issues with the new Smash.All. I made it by doing autoset attributes after adding all the record types (as in the description of the original Smash.All). Is there something more I should have done? I intend to look more closely at their differences when I have the time...

Aside from those two things, I am not aware of any new issues. The issue with Russian is still present but it was already there AFAIK.

Commit:
Maybe fix the new Smash.All

Commit:
Make original BSA loading code work

Commit:
Add setting to ESL flag Smashed Patch

Commit:
Add option to use multiple threads to remove ITPOs

TODO: Clean up this code
TODO: Maybe do the same for smashing records?

Commit:
Further tweak new Smash.All

Commit:
Tweak new Smash.All

Commit:
Attempt at updating Smash.All for FO4

Commit:
Mostly complete Russian translation?

Commit:
Credit Koriss_ for translation

Commit:
Add support for Enderal (SE)

Commit:
Tweaks to processing/merging of arrays and flags

Commit:
Tweak to handling of masters

Check overrides for parent elements and add their masters.
Sort masters according to load order at time of smash.

Commit:
Fix copying new elements from source to dest

Commit:
Don't sort masters

Some of the records were not getting updated properly.

Commit:
New attempt and including parent elements?

Commit:
Only try to update Count fields that are integers

Commit:
Better handling of unions?

Commit:
Tweaks to how elements are copied in

Hopefully this will work better with fewer exceptions.

Commit:
Start calling this v2

Commit:
Fix arrays maybe?

Commit:
Use FullPath instead of Path in debug messages

Commit:
Tweak to picking record masters

It still works wrong for USMP for some reason...

Commit:
Various tweaks to my Smash.All

Commit:
Fix for silently failing to rename smashed patch

I forget to close xedit a lot...

Commit:
Better array alignment

Commit:
Always compare array elements by SortKey?

It exists even if the array is not sorted.
Fixes error from confusion when mods change EditorID of elements.

Commit:
Fix for smashing parents before children

Commit:
Better solution for changing editorid

Commit:
Look at "diffs" relative to all masters.

Commit:
Treat non-existent master arrays as empty

Commit:
Fix for remove not always removing the element

Commit:
Workaround for MasterOrSelf giving wrong master.

Need to figure out what I do wrong to make that happen.
Especially seems to happen with USMP?
This workaround is not perfect, but less broken than before...

Commit:
Fix message indentation

Commit:
List myself as v2 author?

Commit:
More tweaks to my Smash.All

Commit:
v2.0.0-beta+2

Commit:
Fix for smashing the same masters repeatedly.

Commit:
Log format fix

Commit:
Switch to async logging for the tracker.

Now the smashing thread does not get held up by logs.
Mostly helps with speed when debugging is on,
but seems slightly faster normally as well.
TODO: Unify the tracker logs and the mte logs?

Commit:
Support smashing record flags

Commit:
Decide overrides to smash before actually smashing

Commit:
Skip if only override to smash is winning override

Commit:
Sort masters

Commit:
Logging tweaks

Commit:
Tell xedit to not build refs?

Commit:
Smash RACE records

Commit:
Don't leave stuff in the recycle bin

Commit:
Fix ITPO removal

Commit:
Update Smash.All

Commit:
Fix load ordering for "ghost"ed files

Commit:
Most fixes for ghosted files

Commit:
Add smashGhosts setting

Commit:
stuff
